<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>13 Modifying DNA Data | Python for Bioinformatics</title>
  <meta name="description" content="NEOF book for the Python for Bioinformatics workshop" />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="13 Modifying DNA Data | Python for Bioinformatics" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="/figures/NEOF.png" />
  <meta property="og:description" content="NEOF book for the Python for Bioinformatics workshop" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="13 Modifying DNA Data | Python for Bioinformatics" />
  
  <meta name="twitter:description" content="NEOF book for the Python for Bioinformatics workshop" />
  <meta name="twitter:image" content="/figures/NEOF.png" />




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="figures/NEOF_favicon.png" type="image/x-icon" />
<link rel="prev" href="12-DNA_sequence_data.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
<link rel="stylesheet" href="www/webex.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="toc-logo"><a href="https://neof.org.uk/" target="blank"><img src="figures/neof_small_logo.png"></a></li>
<li><a>Python for Bioinformatics</a></li>

<li class="divider"></li>
<li class="part"><span><b>Intro</b></span></li>
<li class="chapter" data-level="1" data-path="01-Python.html"><a href="01-Python.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="02-intro.html"><a href="02-intro.html"><i class="fa fa-check"></i><b>2</b> Python for bioinformatics</a>
<ul>
<li class="chapter" data-level="2.1" data-path="02-intro.html"><a href="02-intro.html#materials-covered-in-the-course"><i class="fa fa-check"></i><b>2.1</b> Materials Covered in the Course</a></li>
<li class="chapter" data-level="2.2" data-path="02-intro.html"><a href="02-intro.html#inputting-commands"><i class="fa fa-check"></i><b>2.2</b> Inputting Commands</a></li>
<li class="chapter" data-level="2.3" data-path="02-intro.html"><a href="02-intro.html#pseudocode"><i class="fa fa-check"></i><b>2.3</b> Pseudocode</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="03_cluster_introduction.html"><a href="03_cluster_introduction.html"><i class="fa fa-check"></i><b>3</b> Cluster Introduction</a>
<ul>
<li class="chapter" data-level="3.1" data-path="03_cluster_introduction.html"><a href="03_cluster_introduction.html#logon-instructions"><i class="fa fa-check"></i><b>3.1</b> Logon Instructions</a></li>
<li class="chapter" data-level="3.2" data-path="03_cluster_introduction.html"><a href="03_cluster_introduction.html#the-terminal-window"><i class="fa fa-check"></i><b>3.2</b> The Terminal Window</a></li>
<li class="chapter" data-level="3.3" data-path="03_cluster_introduction.html"><a href="03_cluster_introduction.html#load-workshop-data"><i class="fa fa-check"></i><b>3.3</b> Load workshop data</a></li>
<li class="chapter" data-level="3.4" data-path="03_cluster_introduction.html"><a href="03_cluster_introduction.html#mamba"><i class="fa fa-check"></i><b>3.4</b> Mamba</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="04-Jupyter.html"><a href="04-Jupyter.html"><i class="fa fa-check"></i><b>4</b> Jupyter</a>
<ul>
<li class="chapter" data-level="4.1" data-path="04-Jupyter.html"><a href="04-Jupyter.html#open-jupyter-notebook"><i class="fa fa-check"></i><b>4.1</b> Open Jupyter-notebook</a></li>
<li class="chapter" data-level="4.2" data-path="04-Jupyter.html"><a href="04-Jupyter.html#changing-directories"><i class="fa fa-check"></i><b>4.2</b> Changing directories</a></li>
<li class="chapter" data-level="4.3" data-path="04-Jupyter.html"><a href="04-Jupyter.html#create-notebook"><i class="fa fa-check"></i><b>4.3</b> Create notebook</a></li>
<li class="chapter" data-level="4.4" data-path="04-Jupyter.html"><a href="04-Jupyter.html#cells-and-code"><i class="fa fa-check"></i><b>4.4</b> Cells and code</a></li>
<li class="chapter" data-level="4.5" data-path="04-Jupyter.html"><a href="04-Jupyter.html#create-new-cells"><i class="fa fa-check"></i><b>4.5</b> Create new cells</a></li>
<li class="chapter" data-level="4.6" data-path="04-Jupyter.html"><a href="04-Jupyter.html#running-code"><i class="fa fa-check"></i><b>4.6</b> Running code</a></li>
<li class="chapter" data-level="4.7" data-path="04-Jupyter.html"><a href="04-Jupyter.html#saving-the-file"><i class="fa fa-check"></i><b>4.7</b> Saving the file</a></li>
<li class="chapter" data-level="4.8" data-path="04-Jupyter.html"><a href="04-Jupyter.html#title-cells-with-markdown"><i class="fa fa-check"></i><b>4.8</b> Title cells with markdown</a></li>
<li class="chapter" data-level="4.9" data-path="04-Jupyter.html"><a href="04-Jupyter.html#close-the-notebook"><i class="fa fa-check"></i><b>4.9</b> Close the notebook</a></li>
<li class="chapter" data-level="4.10" data-path="04-Jupyter.html"><a href="04-Jupyter.html#video-tutorial"><i class="fa fa-check"></i><b>4.10</b> Video tutorial</a></li>
</ul></li>
<li class="part"><span><b>Fundamentals</b></span></li>
<li class="chapter" data-level="5" data-path="05-basic_coding.html"><a href="05-basic_coding.html"><i class="fa fa-check"></i><b>5</b> Scripting in Python</a>
<ul>
<li class="chapter" data-level="5.1" data-path="05-basic_coding.html"><a href="05-basic_coding.html#introduction-to-commands-and-scripts"><i class="fa fa-check"></i><b>5.1</b> Introduction to commands and scripts</a></li>
<li class="chapter" data-level="5.2" data-path="05-basic_coding.html"><a href="05-basic_coding.html#issuing-commands"><i class="fa fa-check"></i><b>5.2</b> Issuing commands</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="05-basic_coding.html"><a href="05-basic_coding.html#print"><i class="fa fa-check"></i><b>5.2.1</b> Print</a></li>
<li class="chapter" data-level="5.2.2" data-path="05-basic_coding.html"><a href="05-basic_coding.html#working-directory"><i class="fa fa-check"></i><b>5.2.2</b> Working directory</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="05-basic_coding.html"><a href="05-basic_coding.html#a-note-on-indentation"><i class="fa fa-check"></i><b>5.3</b> A note on indentation</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="05-basic_coding.html"><a href="05-basic_coding.html#tabs-vs-spaces"><i class="fa fa-check"></i><b>5.3.1</b> Tabs vs spaces</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="05-basic_coding.html"><a href="05-basic_coding.html#scripting-recap-mcqs"><i class="fa fa-check"></i><b>5.4</b> Scripting recap MCQs</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="06-python_fundamentals.html"><a href="06-python_fundamentals.html"><i class="fa fa-check"></i><b>6</b> Python Fundamentals</a>
<ul>
<li class="chapter" data-level="6.1" data-path="06-python_fundamentals.html"><a href="06-python_fundamentals.html#operators"><i class="fa fa-check"></i><b>6.1</b> Operators</a></li>
<li class="chapter" data-level="6.2" data-path="06-python_fundamentals.html"><a href="06-python_fundamentals.html#commentsannotations"><i class="fa fa-check"></i><b>6.2</b> Comments/Annotations</a></li>
<li class="chapter" data-level="6.3" data-path="06-python_fundamentals.html"><a href="06-python_fundamentals.html#variables"><i class="fa fa-check"></i><b>6.3</b> Variables</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="06-python_fundamentals.html"><a href="06-python_fundamentals.html#strings"><i class="fa fa-check"></i><b>6.3.1</b> Strings</a></li>
<li class="chapter" data-level="6.3.2" data-path="06-python_fundamentals.html"><a href="06-python_fundamentals.html#integers"><i class="fa fa-check"></i><b>6.3.2</b> Integers</a></li>
<li class="chapter" data-level="6.3.3" data-path="06-python_fundamentals.html"><a href="06-python_fundamentals.html#floats-floating-point-numbers"><i class="fa fa-check"></i><b>6.3.3</b> Floats (floating point numbers)</a></li>
<li class="chapter" data-level="6.3.4" data-path="06-python_fundamentals.html"><a href="06-python_fundamentals.html#booleans"><i class="fa fa-check"></i><b>6.3.4</b> Booleans</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="06-python_fundamentals.html"><a href="06-python_fundamentals.html#classes-of-variable"><i class="fa fa-check"></i><b>6.4</b> Classes of variable</a></li>
<li class="chapter" data-level="6.5" data-path="06-python_fundamentals.html"><a href="06-python_fundamentals.html#transforming-variables"><i class="fa fa-check"></i><b>6.5</b> Transforming variables</a></li>
<li class="chapter" data-level="6.6" data-path="06-python_fundamentals.html"><a href="06-python_fundamentals.html#why-so-many-code-cells"><i class="fa fa-check"></i><b>6.6</b> Why so many code cells?</a></li>
<li class="chapter" data-level="6.7" data-path="06-python_fundamentals.html"><a href="06-python_fundamentals.html#fundamentals-mcq-recap"><i class="fa fa-check"></i><b>6.7</b> Fundamentals MCQ recap</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="07-functions_and_methods.html"><a href="07-functions_and_methods.html"><i class="fa fa-check"></i><b>7</b> Functions and methods</a>
<ul>
<li class="chapter" data-level="7.1" data-path="07-functions_and_methods.html"><a href="07-functions_and_methods.html#functions"><i class="fa fa-check"></i><b>7.1</b> Functions</a></li>
<li class="chapter" data-level="7.2" data-path="07-functions_and_methods.html"><a href="07-functions_and_methods.html#methods"><i class="fa fa-check"></i><b>7.2</b> Methods</a></li>
<li class="chapter" data-level="7.3" data-path="07-functions_and_methods.html"><a href="07-functions_and_methods.html#functions-and-methods-mcq-recap"><i class="fa fa-check"></i><b>7.3</b> Functions and methods MCQ recap</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="08-data_structures.html"><a href="08-data_structures.html"><i class="fa fa-check"></i><b>8</b> Data Structures</a>
<ul>
<li class="chapter" data-level="8.1" data-path="08-data_structures.html"><a href="08-data_structures.html#lists"><i class="fa fa-check"></i><b>8.1</b> Lists</a></li>
<li class="chapter" data-level="8.2" data-path="08-data_structures.html"><a href="08-data_structures.html#subsetting-data"><i class="fa fa-check"></i><b>8.2</b> Subsetting data</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="08-data_structures.html"><a href="08-data_structures.html#single-values"><i class="fa fa-check"></i><b>8.2.1</b> Single values</a></li>
<li class="chapter" data-level="8.2.2" data-path="08-data_structures.html"><a href="08-data_structures.html#multiple-values"><i class="fa fa-check"></i><b>8.2.2</b> Multiple values</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="08-data_structures.html"><a href="08-data_structures.html#set"><i class="fa fa-check"></i><b>8.3</b> Set</a></li>
<li class="chapter" data-level="8.4" data-path="08-data_structures.html"><a href="08-data_structures.html#convert-a-list-to-a-set"><i class="fa fa-check"></i><b>8.4</b> Convert a list to a set</a></li>
<li class="chapter" data-level="8.5" data-path="08-data_structures.html"><a href="08-data_structures.html#dictionary"><i class="fa fa-check"></i><b>8.5</b> Dictionary</a></li>
<li class="chapter" data-level="8.6" data-path="08-data_structures.html"><a href="08-data_structures.html#data-structures-mcq-recap"><i class="fa fa-check"></i><b>8.6</b> Data structures MCQ recap</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="09-fundamental-exercises.html"><a href="09-fundamental-exercises.html"><i class="fa fa-check"></i><b>9</b> Fundamental exercise</a>
<ul>
<li class="chapter" data-level="9.1" data-path="09-fundamental-exercises.html"><a href="09-fundamental-exercises.html#fundamental-exercise-notebook"><i class="fa fa-check"></i><b>9.1</b> Fundamental exercise notebook</a></li>
<li class="chapter" data-level="9.2" data-path="09-fundamental-exercises.html"><a href="09-fundamental-exercises.html#moth-light-trap-challenges"><i class="fa fa-check"></i><b>9.2</b> Moth light trap challenges</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="09-fundamental-exercises.html"><a href="09-fundamental-exercises.html#challenge-1"><i class="fa fa-check"></i><b>9.2.1</b> Challenge 1</a></li>
<li class="chapter" data-level="9.2.2" data-path="09-fundamental-exercises.html"><a href="09-fundamental-exercises.html#challenge-2"><i class="fa fa-check"></i><b>9.2.2</b> Challenge 2</a></li>
<li class="chapter" data-level="9.2.3" data-path="09-fundamental-exercises.html"><a href="09-fundamental-exercises.html#challenge-3"><i class="fa fa-check"></i><b>9.2.3</b> Challenge 3</a></li>
<li class="chapter" data-level="9.2.4" data-path="09-fundamental-exercises.html"><a href="09-fundamental-exercises.html#challenge-4"><i class="fa fa-check"></i><b>9.2.4</b> Challenge 4</a></li>
<li class="chapter" data-level="9.2.5" data-path="09-fundamental-exercises.html"><a href="09-fundamental-exercises.html#challenge-5"><i class="fa fa-check"></i><b>9.2.5</b> Challenge 5</a></li>
<li class="chapter" data-level="9.2.6" data-path="09-fundamental-exercises.html"><a href="09-fundamental-exercises.html#challenge-6"><i class="fa fa-check"></i><b>9.2.6</b> Challenge 6</a></li>
<li class="chapter" data-level="9.2.7" data-path="09-fundamental-exercises.html"><a href="09-fundamental-exercises.html#challenge-7"><i class="fa fa-check"></i><b>9.2.7</b> Challenge 7</a></li>
<li class="chapter" data-level="9.2.8" data-path="09-fundamental-exercises.html"><a href="09-fundamental-exercises.html#challenge-8"><i class="fa fa-check"></i><b>9.2.8</b> Challenge 8</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="09-fundamental-exercises.html"><a href="09-fundamental-exercises.html#exercise-conclusion"><i class="fa fa-check"></i><b>9.3</b> Exercise conclusion</a></li>
</ul></li>
<li class="part"><span><b>Programmatic coding</b></span></li>
<li class="chapter" data-level="10" data-path="10-conditions.html"><a href="10-conditions.html"><i class="fa fa-check"></i><b>10</b> Python Conditions</a>
<ul>
<li class="chapter" data-level="10.1" data-path="10-conditions.html"><a href="10-conditions.html#logical-conditions"><i class="fa fa-check"></i><b>10.1</b> Logical conditions</a></li>
<li class="chapter" data-level="10.2" data-path="10-conditions.html"><a href="10-conditions.html#if-statements"><i class="fa fa-check"></i><b>10.2</b> If statements</a></li>
<li class="chapter" data-level="10.3" data-path="10-conditions.html"><a href="10-conditions.html#if-else-statements"><i class="fa fa-check"></i><b>10.3</b> If else statements</a>
<ul>
<li class="chapter" data-level="10.3.1" data-path="10-conditions.html"><a href="10-conditions.html#dna-length-comparison"><i class="fa fa-check"></i><b>10.3.1</b> DNA length comparison</a></li>
<li class="chapter" data-level="10.3.2" data-path="10-conditions.html"><a href="10-conditions.html#nucleotide-count-comparison"><i class="fa fa-check"></i><b>10.3.2</b> Nucleotide count comparison</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="10-conditions.html"><a href="10-conditions.html#conditions-reflection"><i class="fa fa-check"></i><b>10.4</b> Conditions reflection</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="11-loops.html"><a href="11-loops.html"><i class="fa fa-check"></i><b>11</b> Loops</a>
<ul>
<li class="chapter" data-level="11.1" data-path="11-loops.html"><a href="11-loops.html#for-loops"><i class="fa fa-check"></i><b>11.1</b> For Loops</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="11-loops.html"><a href="11-loops.html#loop-through-dna-nucleotides"><i class="fa fa-check"></i><b>11.1.1</b> Loop through DNA nucleotides</a></li>
<li class="chapter" data-level="11.1.2" data-path="11-loops.html"><a href="11-loops.html#complement-a-dna-sequence"><i class="fa fa-check"></i><b>11.1.2</b> Complement a DNA sequence</a></li>
<li class="chapter" data-level="11.1.3" data-path="11-loops.html"><a href="11-loops.html#reverse-complement-a-dna-sequence"><i class="fa fa-check"></i><b>11.1.3</b> Reverse-complement a DNA sequence</a></li>
<li class="chapter" data-level="11.1.4" data-path="11-loops.html"><a href="11-loops.html#transcription-challenge"><i class="fa fa-check"></i><b>11.1.4</b> Transcription challenge</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="11-loops.html"><a href="11-loops.html#while-loops"><i class="fa fa-check"></i><b>11.2</b> While loops</a></li>
<li class="chapter" data-level="11.3" data-path="11-loops.html"><a href="11-loops.html#further-control-statements"><i class="fa fa-check"></i><b>11.3</b> Further control statements</a></li>
<li class="chapter" data-level="11.4" data-path="11-loops.html"><a href="11-loops.html#loop-reflection"><i class="fa fa-check"></i><b>11.4</b> Loop reflection</a></li>
</ul></li>
<li class="part"><span><b>Practical python</b></span></li>
<li class="chapter" data-level="12" data-path="12-DNA_sequence_data.html"><a href="12-DNA_sequence_data.html"><i class="fa fa-check"></i><b>12</b> Reading sequence data</a>
<ul>
<li class="chapter" data-level="12.1" data-path="12-DNA_sequence_data.html"><a href="12-DNA_sequence_data.html#seqio"><i class="fa fa-check"></i><b>12.1</b> SeqIO</a></li>
<li class="chapter" data-level="12.2" data-path="12-DNA_sequence_data.html"><a href="12-DNA_sequence_data.html#the-fastq-format"><i class="fa fa-check"></i><b>12.2</b> The FASTQ format</a></li>
<li class="chapter" data-level="12.3" data-path="12-DNA_sequence_data.html"><a href="12-DNA_sequence_data.html#read-a-fastq-file-into-python"><i class="fa fa-check"></i><b>12.3</b> Read a FASTQ file into Python</a>
<ul>
<li class="chapter" data-level="12.3.1" data-path="12-DNA_sequence_data.html"><a href="12-DNA_sequence_data.html#loading-the-python-library"><i class="fa fa-check"></i><b>12.3.1</b> Loading the python library</a></li>
<li class="chapter" data-level="12.3.2" data-path="12-DNA_sequence_data.html"><a href="12-DNA_sequence_data.html#read-a-file-of-sequence-data"><i class="fa fa-check"></i><b>12.3.2</b> Read a file of sequence data</a></li>
<li class="chapter" data-level="12.3.3" data-path="12-DNA_sequence_data.html"><a href="12-DNA_sequence_data.html#fastq-record-elements"><i class="fa fa-check"></i><b>12.3.3</b> Fastq record elements</a></li>
</ul></li>
<li class="chapter" data-level="12.4" data-path="12-DNA_sequence_data.html"><a href="12-DNA_sequence_data.html#reading-and-using-files"><i class="fa fa-check"></i><b>12.4</b> Reading and using files</a></li>
<li class="chapter" data-level="12.5" data-path="12-DNA_sequence_data.html"><a href="12-DNA_sequence_data.html#modifying-reads"><i class="fa fa-check"></i><b>12.5</b> Modifying reads</a></li>
<li class="chapter" data-level="12.6" data-path="12-DNA_sequence_data.html"><a href="12-DNA_sequence_data.html#count-reads"><i class="fa fa-check"></i><b>12.6</b> Count reads</a></li>
<li class="chapter" data-level="12.7" data-path="12-DNA_sequence_data.html"><a href="12-DNA_sequence_data.html#reading-sequence-data-mcq-recap"><i class="fa fa-check"></i><b>12.7</b> Reading sequence data MCQ recap</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="13_DNA_conditions.html"><a href="13_DNA_conditions.html"><i class="fa fa-check"></i><b>13</b> Modifying DNA Data</a>
<ul>
<li class="chapter" data-level="13.1" data-path="13_DNA_conditions.html"><a href="13_DNA_conditions.html#length-filtering"><i class="fa fa-check"></i><b>13.1</b> Length filtering</a></li>
<li class="chapter" data-level="13.2" data-path="13_DNA_conditions.html"><a href="13_DNA_conditions.html#remove-pcr-primers"><i class="fa fa-check"></i><b>13.2</b> Remove PCR primers</a></li>
<li class="chapter" data-level="13.3" data-path="13_DNA_conditions.html"><a href="13_DNA_conditions.html#filter-by-gc"><i class="fa fa-check"></i><b>13.3</b> Filter by GC%</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="13_DNA_conditions.html"><a href="13_DNA_conditions.html#and-logical-operator"><i class="fa fa-check"></i><b>13.3.1</b> and logical operator</a></li>
<li class="chapter" data-level="13.3.2" data-path="13_DNA_conditions.html"><a href="13_DNA_conditions.html#gc-filter"><i class="fa fa-check"></i><b>13.3.2</b> GC filter</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="13_DNA_conditions.html"><a href="13_DNA_conditions.html#recap"><i class="fa fa-check"></i><b>13.4</b> Recap</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Python for Bioinformatics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="modifying-dna-data" class="section level1 hasAnchor" number="13">
<h1><span class="header-section-number">13</span> Modifying DNA Data<a href="13_DNA_conditions.html#modifying-dna-data" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<center>
<img src="figures/checks.jpeg" height="200" /><img src="figures/DNA_icon2.png" height="200" />
</center>
<p>Suppose we want to filter raw sequence reads by some parameter.
For example, we could:</p>
<ul>
<li>Remove reads of a certain length.</li>
<li>Remove sections of a sequence such as PCR primers.</li>
<li>Filter reads by GC%.</li>
</ul>
<p>In this chapter we will use our knowledge of <strong>loops</strong> and <strong>conditions</strong> to perform these tasks. Continue using <strong>"04-Practical_python"</strong> notebook.</p>
<div id="length-filtering" class="section level2 hasAnchor" number="13.1">
<h2><span class="header-section-number">13.1</span> Length filtering<a href="13_DNA_conditions.html#length-filtering" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><img src="figures/comp_filter.png" width="20%" style="display: block; margin: auto;" /></p>
<p>Using what we know, we will create code to remove/filter reads shorter than a specified read-length threshold from a FastQ file.</p>
<div class="sourceCode" id="cb154"><pre class="sourceCode python Rchunk"><code class="sourceCode python"><span id="cb154-1"><a href="13_DNA_conditions.html#cb154-1" tabindex="-1"></a><span class="co">#Read in the fastq data</span></span>
<span id="cb154-2"><a href="13_DNA_conditions.html#cb154-2" tabindex="-1"></a>R_undulata_data <span class="op">=</span> SeqIO.parse(wd <span class="op">+</span> <span class="st">&quot;/Python_course_data/R_undulata_WGS.fastq&quot;</span>, <span class="st">&quot;fastq&quot;</span>)</span>
<span id="cb154-3"><a href="13_DNA_conditions.html#cb154-3" tabindex="-1"></a></span>
<span id="cb154-4"><a href="13_DNA_conditions.html#cb154-4" tabindex="-1"></a><span class="co">#Initialise an empty list to hold the records that we want to keep</span></span>
<span id="cb154-5"><a href="13_DNA_conditions.html#cb154-5" tabindex="-1"></a>wanted_sequences <span class="op">=</span> []</span>
<span id="cb154-6"><a href="13_DNA_conditions.html#cb154-6" tabindex="-1"></a></span>
<span id="cb154-7"><a href="13_DNA_conditions.html#cb154-7" tabindex="-1"></a><span class="co">#Variable for choice of minimum length</span></span>
<span id="cb154-8"><a href="13_DNA_conditions.html#cb154-8" tabindex="-1"></a><span class="co">#in this case we will choose 100</span></span>
<span id="cb154-9"><a href="13_DNA_conditions.html#cb154-9" tabindex="-1"></a>min_len <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb154-10"><a href="13_DNA_conditions.html#cb154-10" tabindex="-1"></a></span>
<span id="cb154-11"><a href="13_DNA_conditions.html#cb154-11" tabindex="-1"></a><span class="co">#Loop through the fastq records</span></span>
<span id="cb154-12"><a href="13_DNA_conditions.html#cb154-12" tabindex="-1"></a><span class="cf">for</span> record <span class="kw">in</span> R_undulata_data:</span>
<span id="cb154-13"><a href="13_DNA_conditions.html#cb154-13" tabindex="-1"></a>  <span class="co">#If condition to determine if sequence length is greater or equal to minimum length</span></span>
<span id="cb154-14"><a href="13_DNA_conditions.html#cb154-14" tabindex="-1"></a>  <span class="cf">if</span> <span class="bu">len</span>(record.seq) <span class="op">&gt;=</span> min_len:</span>
<span id="cb154-15"><a href="13_DNA_conditions.html#cb154-15" tabindex="-1"></a>    <span class="co">#Add the record to the list of records meets minimum length threshold</span></span>
<span id="cb154-16"><a href="13_DNA_conditions.html#cb154-16" tabindex="-1"></a>    wanted_sequences.append(record)</span>
<span id="cb154-17"><a href="13_DNA_conditions.html#cb154-17" tabindex="-1"></a></span>
<span id="cb154-18"><a href="13_DNA_conditions.html#cb154-18" tabindex="-1"></a><span class="co">#Print statement to show number of reads retained after length filtering</span></span>
<span id="cb154-19"><a href="13_DNA_conditions.html#cb154-19" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;Found &quot;</span> <span class="op">+</span> <span class="bu">str</span>(<span class="bu">len</span>(wanted_sequences)) <span class="op">+</span> <span class="st">&quot; reads &gt;= &quot;</span> <span class="op">+</span> <span class="bu">str</span>(min_len) <span class="st">&quot;bp in length.&quot;</span>)</span></code></pre></div>
</div>
<div id="remove-pcr-primers" class="section level2 hasAnchor" number="13.2">
<h2><span class="header-section-number">13.2</span> Remove PCR primers<a href="13_DNA_conditions.html#remove-pcr-primers" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="figures/cut.png" style="width:20.0%" /><img src="figures/DNA_icon.png" />
</center>
<p>Often you may want to check your sequence file for PCR primers, and remove them when they are found.
Using a <strong>for</strong> <strong>loop</strong> to move through the sequence file, we can check each DNA sequence read for the presence of a pre-defined primer sequence.
If it is present, we'll remove it.</p>
<p>We are going to check for the presence of the 13 nucleotide PCR primer "ACACTGCTGATCG" at the start of the reads using the <code>startswith()</code> <strong>method</strong>.
This <strong>method</strong> checks whether a <strong>string</strong> starts with a particular sub-string and returns <strong>True</strong> or <strong>False</strong>.
It is very useful for checking for primer sequences in our data.</p>
<p>The input for this primer removal step will be the 'wanted_sequences' <strong>variable</strong> from the read filtering code (those of &gt;=100bp in length).</p>
<p>For demonstration purposes we will first run code that will tell us the number of sequences which have the PCR sequence present.</p>
<p><strong>Note</strong>: Below is the primer in triplet bases for ease of reading. Ensure you type it as one string with no spaces.</p>
<p><strong>ACA CTG CTG ATC G</strong></p>
<div class="sourceCode" id="cb155"><pre class="sourceCode python Rchunk"><code class="sourceCode python"><span id="cb155-1"><a href="13_DNA_conditions.html#cb155-1" tabindex="-1"></a><span class="co">#Assign PCR sequence variable</span></span>
<span id="cb155-2"><a href="13_DNA_conditions.html#cb155-2" tabindex="-1"></a>pcr_seq <span class="op">=</span> <span class="st">&quot;ACACTGCTGATCG&quot;</span></span>
<span id="cb155-3"><a href="13_DNA_conditions.html#cb155-3" tabindex="-1"></a><span class="co">#Initialise variable to sum number of records with PCR primer present</span></span>
<span id="cb155-4"><a href="13_DNA_conditions.html#cb155-4" tabindex="-1"></a>pcr_records <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb155-5"><a href="13_DNA_conditions.html#cb155-5" tabindex="-1"></a></span>
<span id="cb155-6"><a href="13_DNA_conditions.html#cb155-6" tabindex="-1"></a><span class="co">#Loop through the sequence records</span></span>
<span id="cb155-7"><a href="13_DNA_conditions.html#cb155-7" tabindex="-1"></a><span class="cf">for</span> record <span class="kw">in</span> wanted_sequences:</span>
<span id="cb155-8"><a href="13_DNA_conditions.html#cb155-8" tabindex="-1"></a>  <span class="co">#Check whether the sequence starts with the primer sequence</span></span>
<span id="cb155-9"><a href="13_DNA_conditions.html#cb155-9" tabindex="-1"></a>  <span class="cf">if</span> record.seq.startswith(<span class="st">&quot;ACACTGCTGATCG&quot;</span>):</span>
<span id="cb155-10"><a href="13_DNA_conditions.html#cb155-10" tabindex="-1"></a>    pcr_records <span class="op">=</span> pcr_records <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb155-11"><a href="13_DNA_conditions.html#cb155-11" tabindex="-1"></a></span>
<span id="cb155-12"><a href="13_DNA_conditions.html#cb155-12" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;Number of sequences with the PCR sequence present: &quot;</span> <span class="op">+</span> <span class="bu">str</span>(pcr_records))</span></code></pre></div>
<p>We have found that there are lots of sequences with the primer sequence present.
Now that we have identified their presence, let's trim them off the start of the reads.</p>
<p>Create and run the below code in a new <strong>cell</strong>.
This will create a new <strong>variable</strong> "trimmed_sequences".
This will contain records that have been:</p>
<ul>
<li>Filtered to remove sequences shorter than 100bp.</li>
<li>Trimmed to remove the PCR sequence ("ACACTGCTGATCG").
<ul>
<li>However, we also want to keep untrimmed sequences that did not contain the PCR sequence.</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb156"><pre class="sourceCode python Rchunk"><code class="sourceCode python"><span id="cb156-1"><a href="13_DNA_conditions.html#cb156-1" tabindex="-1"></a><span class="co">#Define a list to hold the trimmed sequences</span></span>
<span id="cb156-2"><a href="13_DNA_conditions.html#cb156-2" tabindex="-1"></a>trimmed_sequences <span class="op">=</span> []</span>
<span id="cb156-3"><a href="13_DNA_conditions.html#cb156-3" tabindex="-1"></a><span class="co">#Assign PCR sequence variable</span></span>
<span id="cb156-4"><a href="13_DNA_conditions.html#cb156-4" tabindex="-1"></a>pcr_seq <span class="op">=</span> <span class="st">&quot;ACACTGCTGATCG&quot;</span></span>
<span id="cb156-5"><a href="13_DNA_conditions.html#cb156-5" tabindex="-1"></a><span class="co">#Initialise variable to sum number of records with PCR primer present</span></span>
<span id="cb156-6"><a href="13_DNA_conditions.html#cb156-6" tabindex="-1"></a>pcr_records <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb156-7"><a href="13_DNA_conditions.html#cb156-7" tabindex="-1"></a></span>
<span id="cb156-8"><a href="13_DNA_conditions.html#cb156-8" tabindex="-1"></a><span class="co">#Loop through the sequence records</span></span>
<span id="cb156-9"><a href="13_DNA_conditions.html#cb156-9" tabindex="-1"></a><span class="cf">for</span> record <span class="kw">in</span> wanted_sequences:</span>
<span id="cb156-10"><a href="13_DNA_conditions.html#cb156-10" tabindex="-1"></a>  <span class="co">#Check whether the sequence starts with the primer sequence</span></span>
<span id="cb156-11"><a href="13_DNA_conditions.html#cb156-11" tabindex="-1"></a>  <span class="cf">if</span> record.seq.startswith(<span class="st">&quot;ACACTGCTGATCG&quot;</span>):</span>
<span id="cb156-12"><a href="13_DNA_conditions.html#cb156-12" tabindex="-1"></a>    <span class="co">#Subset read from position 14 onwards and add to trimmed_sequences</span></span>
<span id="cb156-13"><a href="13_DNA_conditions.html#cb156-13" tabindex="-1"></a>    <span class="co">#This will remove the first 13 bases, the PCR sequence</span></span>
<span id="cb156-14"><a href="13_DNA_conditions.html#cb156-14" tabindex="-1"></a>    trimmed_sequences.append(record[<span class="dv">13</span>:])</span>
<span id="cb156-15"><a href="13_DNA_conditions.html#cb156-15" tabindex="-1"></a>    pcr_records <span class="op">=</span> pcr_records <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb156-16"><a href="13_DNA_conditions.html#cb156-16" tabindex="-1"></a>  <span class="cf">else</span>:</span>
<span id="cb156-17"><a href="13_DNA_conditions.html#cb156-17" tabindex="-1"></a>    <span class="co">#Add untrimmed sequence to trimmed_sequences</span></span>
<span id="cb156-18"><a href="13_DNA_conditions.html#cb156-18" tabindex="-1"></a>    <span class="co">#this is because there is no PCR sequence present to trim off</span></span>
<span id="cb156-19"><a href="13_DNA_conditions.html#cb156-19" tabindex="-1"></a>    trimmed_sequences.append(record)</span>
<span id="cb156-20"><a href="13_DNA_conditions.html#cb156-20" tabindex="-1"></a></span>
<span id="cb156-21"><a href="13_DNA_conditions.html#cb156-21" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;Number of sequences with the PCR sequence present: &quot;</span> <span class="op">+</span> <span class="bu">str</span>(pcr_records))</span>
<span id="cb156-22"><a href="13_DNA_conditions.html#cb156-22" tabindex="-1"></a></span>
<span id="cb156-23"><a href="13_DNA_conditions.html#cb156-23" tabindex="-1"></a></span>
<span id="cb156-24"><a href="13_DNA_conditions.html#cb156-24" tabindex="-1"></a><span class="co"># define a list to hold the trimmed sequences</span></span>
<span id="cb156-25"><a href="13_DNA_conditions.html#cb156-25" tabindex="-1"></a>trimmed_sequences <span class="op">=</span> []</span>
<span id="cb156-26"><a href="13_DNA_conditions.html#cb156-26" tabindex="-1"></a><span class="co">#Start counter to count number of reads with primer</span></span>
<span id="cb156-27"><a href="13_DNA_conditions.html#cb156-27" tabindex="-1"></a>counter <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb156-28"><a href="13_DNA_conditions.html#cb156-28" tabindex="-1"></a></span>
<span id="cb156-29"><a href="13_DNA_conditions.html#cb156-29" tabindex="-1"></a><span class="co"># loop through the sequence records</span></span>
<span id="cb156-30"><a href="13_DNA_conditions.html#cb156-30" tabindex="-1"></a><span class="cf">for</span> record <span class="kw">in</span> wanted_sequences:</span>
<span id="cb156-31"><a href="13_DNA_conditions.html#cb156-31" tabindex="-1"></a>  </span>
<span id="cb156-32"><a href="13_DNA_conditions.html#cb156-32" tabindex="-1"></a>  <span class="co"># check whether the sequence starts with the primer sequence</span></span>
<span id="cb156-33"><a href="13_DNA_conditions.html#cb156-33" tabindex="-1"></a>  <span class="cf">if</span> record.seq.startswith(<span class="st">&quot;ACACTGCTGATCG&quot;</span>):</span>
<span id="cb156-34"><a href="13_DNA_conditions.html#cb156-34" tabindex="-1"></a>    <span class="co">#print(record.seq)</span></span>
<span id="cb156-35"><a href="13_DNA_conditions.html#cb156-35" tabindex="-1"></a>    </span>
<span id="cb156-36"><a href="13_DNA_conditions.html#cb156-36" tabindex="-1"></a>    <span class="co"># if it does, subset the read from position 14 onwards </span></span>
<span id="cb156-37"><a href="13_DNA_conditions.html#cb156-37" tabindex="-1"></a>    <span class="co"># and store in trimmed_sequences</span></span>
<span id="cb156-38"><a href="13_DNA_conditions.html#cb156-38" tabindex="-1"></a>    trimmed_sequences.append(record[<span class="dv">13</span>:])</span>
<span id="cb156-39"><a href="13_DNA_conditions.html#cb156-39" tabindex="-1"></a>    counter <span class="op">=</span> counter <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb156-40"><a href="13_DNA_conditions.html#cb156-40" tabindex="-1"></a>  </span>
<span id="cb156-41"><a href="13_DNA_conditions.html#cb156-41" tabindex="-1"></a>  <span class="cf">else</span>:</span>
<span id="cb156-42"><a href="13_DNA_conditions.html#cb156-42" tabindex="-1"></a>    <span class="co"># if it doesn&#39;t have the primer, </span></span>
<span id="cb156-43"><a href="13_DNA_conditions.html#cb156-43" tabindex="-1"></a>    <span class="co"># retain the whole read and store in trimmed_sequences</span></span>
<span id="cb156-44"><a href="13_DNA_conditions.html#cb156-44" tabindex="-1"></a>    trimmed_sequences.append(record)</span>
<span id="cb156-45"><a href="13_DNA_conditions.html#cb156-45" tabindex="-1"></a>    </span>
<span id="cb156-46"><a href="13_DNA_conditions.html#cb156-46" tabindex="-1"></a>number_sequences_trimmed <span class="op">=</span> counter</span>
<span id="cb156-47"><a href="13_DNA_conditions.html#cb156-47" tabindex="-1"></a></span>
<span id="cb156-48"><a href="13_DNA_conditions.html#cb156-48" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;Found and trimmed &quot;</span> <span class="op">+</span> <span class="bu">str</span>(number_sequences_trimmed) <span class="op">+</span> <span class="op">\</span></span>
<span id="cb156-49"><a href="13_DNA_conditions.html#cb156-49" tabindex="-1"></a><span class="st">&quot; reads with PCR primer present at start.&quot;</span>)</span></code></pre></div>
<p>Save and run the code one more time to see your three lines of information printed to screen.</p>
</div>
<div id="filter-by-gc" class="section level2 hasAnchor" number="13.3">
<h2><span class="header-section-number">13.3</span> Filter by GC%<a href="13_DNA_conditions.html#filter-by-gc" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The next bit of coding will retain reads with a GC% between 30% and 70%. This is not normally something you would do but instead we've included it for demonstration purposes.</p>
<div id="and-logical-operator" class="section level3 hasAnchor" number="13.3.1">
<h3><span class="header-section-number">13.3.1</span> and logical operator<a href="13_DNA_conditions.html#and-logical-operator" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>For this we will use the <code>and</code> <strong>logical operator</strong>. This allows us to test multiple comparisons <code>and</code> <strong>if</strong> they are all <code>True</code> the <strong>condition</strong> will return the output. Lets show a few examples.</p>
<p>First create a few demonstration variables.</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode python Rchunk"><code class="sourceCode python"><span id="cb157-1"><a href="13_DNA_conditions.html#cb157-1" tabindex="-1"></a>num_list <span class="op">=</span> [<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">7</span>,<span class="dv">8</span>,<span class="dv">9</span>]</span>
<span id="cb157-2"><a href="13_DNA_conditions.html#cb157-2" tabindex="-1"></a>string_list <span class="op">=</span> [<span class="st">&quot;one&quot;</span>,<span class="st">&quot;two&quot;</span>,<span class="st">&quot;three&quot;</span>,<span class="st">&quot;four&quot;</span>,<span class="st">&quot;five&quot;</span>,<span class="st">&quot;six&quot;</span>,<span class="st">&quot;seven&quot;</span>,<span class="st">&quot;eight&quot;</span>,<span class="st">&quot;nine&quot;</span>]</span></code></pre></div>
<p>Print numbers between 3 and 7.</p>
<div class="sourceCode" id="cb158"><pre class="sourceCode python Rchunk"><code class="sourceCode python"><span id="cb158-1"><a href="13_DNA_conditions.html#cb158-1" tabindex="-1"></a><span class="co">#Loop through numbers</span></span>
<span id="cb158-2"><a href="13_DNA_conditions.html#cb158-2" tabindex="-1"></a><span class="cf">for</span> num <span class="kw">in</span> num_list:</span>
<span id="cb158-3"><a href="13_DNA_conditions.html#cb158-3" tabindex="-1"></a>  <span class="co">#If number is greater than 3 and less than 7</span></span>
<span id="cb158-4"><a href="13_DNA_conditions.html#cb158-4" tabindex="-1"></a>  <span class="cf">if</span> num <span class="op">&gt;</span> <span class="dv">3</span> <span class="kw">and</span> num <span class="op">&lt;</span> <span class="dv">7</span>:</span>
<span id="cb158-5"><a href="13_DNA_conditions.html#cb158-5" tabindex="-1"></a>    <span class="co">#Print number</span></span>
<span id="cb158-6"><a href="13_DNA_conditions.html#cb158-6" tabindex="-1"></a>    <span class="bu">print</span>(num)</span></code></pre></div>
<p>Print numbers divisible by 2 that are greater than 5.</p>
<div class="sourceCode" id="cb159"><pre class="sourceCode python Rchunk"><code class="sourceCode python"><span id="cb159-1"><a href="13_DNA_conditions.html#cb159-1" tabindex="-1"></a><span class="co">#Loop through numbers</span></span>
<span id="cb159-2"><a href="13_DNA_conditions.html#cb159-2" tabindex="-1"></a><span class="cf">for</span> num <span class="kw">in</span> num_list:</span>
<span id="cb159-3"><a href="13_DNA_conditions.html#cb159-3" tabindex="-1"></a>  <span class="co">#If number is divisible by 2 and greater than 5</span></span>
<span id="cb159-4"><a href="13_DNA_conditions.html#cb159-4" tabindex="-1"></a>  <span class="cf">if</span> num<span class="op">%</span><span class="dv">2</span> <span class="op">==</span> <span class="dv">0</span> <span class="kw">and</span> num <span class="op">&gt;</span> <span class="dv">5</span>:</span>
<span id="cb159-5"><a href="13_DNA_conditions.html#cb159-5" tabindex="-1"></a>    <span class="co">#Print number</span></span>
<span id="cb159-6"><a href="13_DNA_conditions.html#cb159-6" tabindex="-1"></a>    <span class="bu">print</span>(num)</span></code></pre></div>
<p>Print strings longer than three characters and has at least on letter <strong>e</strong>.</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode python Rchunk"><code class="sourceCode python"><span id="cb160-1"><a href="13_DNA_conditions.html#cb160-1" tabindex="-1"></a><span class="co">#Loop through strings</span></span>
<span id="cb160-2"><a href="13_DNA_conditions.html#cb160-2" tabindex="-1"></a><span class="cf">for</span> string <span class="kw">in</span> string_list:</span>
<span id="cb160-3"><a href="13_DNA_conditions.html#cb160-3" tabindex="-1"></a>  <span class="co">#If string length is greater than 3 and the &quot;e&quot; count is greater than 0</span></span>
<span id="cb160-4"><a href="13_DNA_conditions.html#cb160-4" tabindex="-1"></a>  <span class="cf">if</span> <span class="bu">len</span>(string) <span class="op">&gt;</span> <span class="dv">3</span> <span class="kw">and</span> string.count(<span class="st">&quot;e&quot;</span>) <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb160-5"><a href="13_DNA_conditions.html#cb160-5" tabindex="-1"></a>    <span class="co">#Print the string</span></span>
<span id="cb160-6"><a href="13_DNA_conditions.html#cb160-6" tabindex="-1"></a>    <span class="bu">print</span>(string)</span></code></pre></div>
<p>Print strings shorter or equal to 3 characters and that have no <strong>e</strong>.</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode python Rchunk"><code class="sourceCode python"><span id="cb161-1"><a href="13_DNA_conditions.html#cb161-1" tabindex="-1"></a><span class="co">#Loop through strings</span></span>
<span id="cb161-2"><a href="13_DNA_conditions.html#cb161-2" tabindex="-1"></a><span class="cf">for</span> string <span class="kw">in</span> string_list:</span>
<span id="cb161-3"><a href="13_DNA_conditions.html#cb161-3" tabindex="-1"></a>  <span class="co">#If string length is less than or equal to 3 and the &quot;e&quot; count is 0</span></span>
<span id="cb161-4"><a href="13_DNA_conditions.html#cb161-4" tabindex="-1"></a>  <span class="cf">if</span> <span class="bu">len</span>(string) <span class="op">&lt;=</span> <span class="dv">3</span> <span class="kw">and</span> string.count(<span class="st">&quot;e&quot;</span>) <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="cb161-5"><a href="13_DNA_conditions.html#cb161-5" tabindex="-1"></a>    <span class="co">#Print the string</span></span>
<span id="cb161-6"><a href="13_DNA_conditions.html#cb161-6" tabindex="-1"></a>    <span class="bu">print</span>(string)</span></code></pre></div>
<p>In the above examples we have only used one <code>and</code> <strong>operator</strong>. You can use multiple but of course the more you use the more complex your <strong>statement</strong> becomes.</p>
</div>
<div id="gc-filter" class="section level3 hasAnchor" number="13.3.2">
<h3><span class="header-section-number">13.3.2</span> GC filter<a href="13_DNA_conditions.html#gc-filter" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>With the <code>and</code> <strong>logical operator</strong> we can filter by a GC% range.</p>
<div class="sourceCode" id="cb162"><pre class="sourceCode python Rchunk"><code class="sourceCode python"><span id="cb162-1"><a href="13_DNA_conditions.html#cb162-1" tabindex="-1"></a><span class="co">#Read in the fastq data</span></span>
<span id="cb162-2"><a href="13_DNA_conditions.html#cb162-2" tabindex="-1"></a>R_undulata_data <span class="op">=</span> SeqIO.parse(wd <span class="op">+</span> <span class="st">&quot;/Python_course_data/R_undulata_WGS.fastq&quot;</span>, <span class="st">&quot;fastq&quot;</span>)</span>
<span id="cb162-3"><a href="13_DNA_conditions.html#cb162-3" tabindex="-1"></a></span>
<span id="cb162-4"><a href="13_DNA_conditions.html#cb162-4" tabindex="-1"></a><span class="co">#Initialise an empty list to hold the records that we want to keep</span></span>
<span id="cb162-5"><a href="13_DNA_conditions.html#cb162-5" tabindex="-1"></a>gc_sequences <span class="op">=</span> []</span>
<span id="cb162-6"><a href="13_DNA_conditions.html#cb162-6" tabindex="-1"></a></span>
<span id="cb162-7"><a href="13_DNA_conditions.html#cb162-7" tabindex="-1"></a><span class="co">#Variable for choice of min and max gc%</span></span>
<span id="cb162-8"><a href="13_DNA_conditions.html#cb162-8" tabindex="-1"></a>min_gc <span class="op">=</span> <span class="dv">40</span></span>
<span id="cb162-9"><a href="13_DNA_conditions.html#cb162-9" tabindex="-1"></a>max_gc <span class="op">=</span> <span class="dv">60</span></span>
<span id="cb162-10"><a href="13_DNA_conditions.html#cb162-10" tabindex="-1"></a></span>
<span id="cb162-11"><a href="13_DNA_conditions.html#cb162-11" tabindex="-1"></a><span class="co">#Loop through the fastq records</span></span>
<span id="cb162-12"><a href="13_DNA_conditions.html#cb162-12" tabindex="-1"></a><span class="cf">for</span> record <span class="kw">in</span> R_undulata_data:</span>
<span id="cb162-13"><a href="13_DNA_conditions.html#cb162-13" tabindex="-1"></a>  <span class="co">#Calculate GC%</span></span>
<span id="cb162-14"><a href="13_DNA_conditions.html#cb162-14" tabindex="-1"></a>  gc_count <span class="op">=</span> record.seq.upper().count(<span class="st">&quot;G&quot;</span>) <span class="op">+</span> record.seq.upper().count(<span class="st">&quot;C&quot;</span>)</span>
<span id="cb162-15"><a href="13_DNA_conditions.html#cb162-15" tabindex="-1"></a>  gc_perc <span class="op">=</span> (gc_count <span class="op">/</span> <span class="bu">len</span>(record.seq)) <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb162-16"><a href="13_DNA_conditions.html#cb162-16" tabindex="-1"></a>  <span class="co">#If GC percent is greater than min and less than max</span></span>
<span id="cb162-17"><a href="13_DNA_conditions.html#cb162-17" tabindex="-1"></a>  <span class="cf">if</span> gc_perc <span class="op">&gt;</span> min_gc <span class="kw">and</span> gc_perc <span class="op">&lt;</span> max_gc:</span>
<span id="cb162-18"><a href="13_DNA_conditions.html#cb162-18" tabindex="-1"></a>    <span class="co">#Add the record to the list of records</span></span>
<span id="cb162-19"><a href="13_DNA_conditions.html#cb162-19" tabindex="-1"></a>    gc_sequences.append(record)</span>
<span id="cb162-20"><a href="13_DNA_conditions.html#cb162-20" tabindex="-1"></a></span>
<span id="cb162-21"><a href="13_DNA_conditions.html#cb162-21" tabindex="-1"></a><span class="co">#Print statement to show number of reads retained after length filtering</span></span>
<span id="cb162-22"><a href="13_DNA_conditions.html#cb162-22" tabindex="-1"></a><span class="bu">print</span>(<span class="st">&quot;Found &quot;</span> <span class="op">+</span> <span class="bu">str</span>(<span class="bu">len</span>(gc_sequences)) <span class="op">+</span> <span class="st">&quot; reads with GC</span><span class="sc">% r</span><span class="st">ange between &quot;</span> <span class="op">+</span> </span>
<span id="cb162-23"><a href="13_DNA_conditions.html#cb162-23" tabindex="-1"></a>  <span class="bu">str</span>(min_gc) <span class="op">+</span> <span class="st">&quot;% and &quot;</span> <span class="op">+</span> <span class="bu">str</span>(max_gc) <span class="op">+</span> <span class="st">&quot;%.&quot;</span>)</span></code></pre></div>
<p>Ensure you understand the above code before continuing.
Please ask a demonstrator if you are unsure.</p>
</div>
</div>
<div id="recap" class="section level2 hasAnchor" number="13.4">
<h2><span class="header-section-number">13.4</span> Recap<a href="13_DNA_conditions.html#recap" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><img src="figures/recap.webp" width="20%" style="display: block; margin: auto;" /></p>
<p>Using our knowledge of reading FASTQ files, looping through the data, and using conditions to check certain characteristics of the data we have been able to filter out short-reads and also remove primer contamination. We are starting to build a pipeline!</p>
<p>We will learn to write data out and store it in a new file in the next chapter.</p>

</div>
</div>
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  var t = document.getElementsByClassName("webex-total_correct");
  for (var i = 0; i < t.length; i++) {
    p = t[i].parentElement;
    var correct = p.getElementsByClassName("webex-correct").length;
    var solvemes = p.getElementsByClassName("webex-solveme").length;
    var radiogroups = p.getElementsByClassName("webex-radiogroup").length;
    var selects = p.getElementsByClassName("webex-select").length;

    t[i].innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");

  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* check answers */
check_func = function() {
  console.log("webex: check answers");

  var cl = this.parentElement.classList;
  if (cl.contains('unchecked')) {
    cl.remove("unchecked");
    this.innerHTML = "Hide Answers";
  } else {
    cl.add("unchecked");
    this.innerHTML = "Show Answers";
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");

  var cl = this.classList

  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;

  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }

  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

window.onload = function() {
  console.log("webex onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  var check_sections = document.getElementsByClassName("webex-check");
  console.log("check:", check_sections.length);
  for (var i = 0; i < check_sections.length; i++) {
    check_sections[i].classList.add("unchecked");

    let btn = document.createElement("button");
    btn.innerHTML = "Show Answers";
    btn.classList.add("webex-check-button");
    btn.onclick = check_func;
    check_sections[i].appendChild(btn);

    let spn = document.createElement("span");
    spn.classList.add("webex-total_correct");
    check_sections[i].appendChild(spn);
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;

    solveme[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }

  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
    selects[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  update_total_correct();
}

</script>
            </section>

          </div>
        </div>
      </div>
<a href="12-DNA_sequence_data.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

</body>

</html>
