<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>15 begin the while loop | Python for Bioinformatics</title>
  <meta name="description" content="15 begin the while loop | Python for Bioinformatics" />
  <meta name="generator" content="bookdown 0.27 and GitBook 2.6.7" />

  <meta property="og:title" content="15 begin the while loop | Python for Bioinformatics" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="15 begin the while loop | Python for Bioinformatics" />
  
  
  

<meta name="author" content="Graeme Fox" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="initialise-a-variable-to-add-the-daily-temps-together.html"/>
<link rel="next" href="save-the-number-of-raw-sequences-for-use-later.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>




</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="python-for-bioinformatics.html"><a href="python-for-bioinformatics.html"><i class="fa fa-check"></i><b>2</b> Python for bioinformatics</a></li>
<li class="chapter" data-level="3" data-path="cluster-introduction.html"><a href="cluster-introduction.html"><i class="fa fa-check"></i><b>3</b> Cluster Introduction</a>
<ul>
<li class="chapter" data-level="3.1" data-path="cluster-introduction.html"><a href="cluster-introduction.html#logon-instructions"><i class="fa fa-check"></i><b>3.1</b> Logon Instructions</a></li>
<li class="chapter" data-level="3.2" data-path="cluster-introduction.html"><a href="cluster-introduction.html#the-terminal-window"><i class="fa fa-check"></i><b>3.2</b> The Terminal Window</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="scripting-in-python.html"><a href="scripting-in-python.html"><i class="fa fa-check"></i><b>4</b> Scripting in Python</a>
<ul>
<li class="chapter" data-level="4.1" data-path="scripting-in-python.html"><a href="scripting-in-python.html#introduction-to-commands-and-scripts"><i class="fa fa-check"></i><b>4.1</b> Introduction to commands and scripts</a></li>
<li class="chapter" data-level="4.2" data-path="scripting-in-python.html"><a href="scripting-in-python.html#issuing-commands-directly-to-the-interpretator"><i class="fa fa-check"></i><b>4.2</b> Issuing commands directly to the interpretator</a></li>
<li class="chapter" data-level="4.3" data-path="scripting-in-python.html"><a href="scripting-in-python.html#a-note-on-indentation"><i class="fa fa-check"></i><b>4.3</b> A note on indentation</a></li>
<li class="chapter" data-level="4.4" data-path="scripting-in-python.html"><a href="scripting-in-python.html#tabs-vs-spaces"><i class="fa fa-check"></i><b>4.4</b> Tabs vs spaces</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="python-fundamentals.html"><a href="python-fundamentals.html"><i class="fa fa-check"></i><b>5</b> Python Fundamentals</a>
<ul>
<li class="chapter" data-level="5.1" data-path="python-fundamentals.html"><a href="python-fundamentals.html#operators"><i class="fa fa-check"></i><b>5.1</b> Operators</a></li>
<li class="chapter" data-level="5.2" data-path="python-fundamentals.html"><a href="python-fundamentals.html#comments"><i class="fa fa-check"></i><b>5.2</b> Comments</a></li>
<li class="chapter" data-level="5.3" data-path="python-fundamentals.html"><a href="python-fundamentals.html#variables"><i class="fa fa-check"></i><b>5.3</b> Variables</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="python-fundamentals.html"><a href="python-fundamentals.html#strings"><i class="fa fa-check"></i><b>5.3.1</b> Strings</a></li>
<li class="chapter" data-level="5.3.2" data-path="python-fundamentals.html"><a href="python-fundamentals.html#integers"><i class="fa fa-check"></i><b>5.3.2</b> Integers</a></li>
<li class="chapter" data-level="5.3.3" data-path="python-fundamentals.html"><a href="python-fundamentals.html#floats-floating-point-numbers"><i class="fa fa-check"></i><b>5.3.3</b> Floats (floating point numbers)</a></li>
<li class="chapter" data-level="5.3.4" data-path="python-fundamentals.html"><a href="python-fundamentals.html#booleans"><i class="fa fa-check"></i><b>5.3.4</b> Booleans</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="python-fundamentals.html"><a href="python-fundamentals.html#classes-of-variable"><i class="fa fa-check"></i><b>5.4</b> Classes of variable</a></li>
<li class="chapter" data-level="5.5" data-path="python-fundamentals.html"><a href="python-fundamentals.html#transforming-variables"><i class="fa fa-check"></i><b>5.5</b> Transforming variables</a></li>
<li class="chapter" data-level="5.6" data-path="python-fundamentals.html"><a href="python-fundamentals.html#functions-and-methods"><i class="fa fa-check"></i><b>5.6</b> Functions and methods</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="python-fundamentals.html"><a href="python-fundamentals.html#running-a-function"><i class="fa fa-check"></i><b>5.6.1</b> Running a function</a></li>
<li class="chapter" data-level="5.6.2" data-path="python-fundamentals.html"><a href="python-fundamentals.html#running-a-method"><i class="fa fa-check"></i><b>5.6.2</b> Running a method</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="get-the-fifth-value.html"><a href="get-the-fifth-value.html"><i class="fa fa-check"></i><b>6</b> Get the fifth value:</a></li>
<li class="chapter" data-level="7" data-path="get-the-last-value-wihtout-knowing-the-length-of-the-list.html"><a href="get-the-last-value-wihtout-knowing-the-length-of-the-list.html"><i class="fa fa-check"></i><b>7</b> Get the last value (wihtout knowing the length of the list)</a></li>
<li class="chapter" data-level="8" data-path="even-though-pygm-and-pgm2-were-each-added-twice.html"><a href="even-though-pygm-and-pgm2-were-each-added-twice.html"><i class="fa fa-check"></i><b>8</b> even though Pygm and Pgm2 were each added twice,</a></li>
<li class="chapter" data-level="9" data-path="they-only-appear-once-in-the-set.html"><a href="they-only-appear-once-in-the-set.html"><i class="fa fa-check"></i><b>9</b> they only appear once in the set</a></li>
<li class="chapter" data-level="10" data-path="add-another-gene-identifier.html"><a href="add-another-gene-identifier.html"><i class="fa fa-check"></i><b>10</b> Add another gene identifier</a></li>
<li class="chapter" data-level="11" data-path="redefine-the-list-since-we-have-moved-into-vscode..html"><a href="redefine-the-list-since-we-have-moved-into-vscode..html"><i class="fa fa-check"></i><b>11</b> redefine the list since we have moved into VSCode.</a></li>
<li class="chapter" data-level="12" data-path="loop-through-each-value-in-the-list.html"><a href="loop-through-each-value-in-the-list.html"><i class="fa fa-check"></i><b>12</b> loop through each value in the list</a></li>
<li class="chapter" data-level="13" data-path="initialise-the-counter-value.html"><a href="initialise-the-counter-value.html"><i class="fa fa-check"></i><b>13</b> initialise the counter value</a></li>
<li class="chapter" data-level="14" data-path="initialise-a-variable-to-add-the-daily-temps-together.html"><a href="initialise-a-variable-to-add-the-daily-temps-together.html"><i class="fa fa-check"></i><b>14</b> initialise a variable to add the daily temps together</a></li>
<li class="chapter" data-level="15" data-path="begin-the-while-loop.html"><a href="begin-the-while-loop.html"><i class="fa fa-check"></i><b>15</b> begin the while loop</a>
<ul>
<li class="chapter" data-level="15.1" data-path="begin-the-while-loop.html"><a href="begin-the-while-loop.html#start-a-counter-to-track-how-many-reads"><i class="fa fa-check"></i><b>15.1</b> start a counter to track how many reads</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="save-the-number-of-raw-sequences-for-use-later.html"><a href="save-the-number-of-raw-sequences-for-use-later.html"><i class="fa fa-check"></i><b>16</b> save the number of raw sequences for use later</a></li>
<li class="chapter" data-level="17" data-path="initialise-an-empty-list-to-hold-the-records-that-we-want-to-keep.html"><a href="initialise-an-empty-list-to-hold-the-records-that-we-want-to-keep.html"><i class="fa fa-check"></i><b>17</b> initialise an empty list to hold the records that we want to keep</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Python for Bioinformatics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="begin-the-while-loop" class="section level1 hasAnchor" number="15">
<h1><span class="header-section-number">15</span> begin the while loop<a href="begin-the-while-loop.html#begin-the-while-loop" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>while counter &lt; 5:
# add the current day peak temperature
temp_running_total = temp_running_total + peak_day_temp_C[counter]
# increase the counter
counter = counter + 1
print(“The average weekday temperature is” + str(temp_running_total/5) +” C.”)</p>
<pre><code>
The *while* statement depends on the value of the counter variable, which starts at zero and is increased by one with each iteration of the loop. The loop will continue to run until the counter has the value of five, at which point the loop stops and the final print statement is performed giving the average weekday temperature.

&lt;b&gt;Question:&lt;/b&gt; What would happen if the &#39;counter = counter + 1&#39; line was not present in the loop above?

## Further control statements
You can control loops further using break, continue and pass statements although we are not covering these in this tutorial as they are much less common. You can find good resources covering Python control statements [online](https://www.tutorialspoint.com/python/python_loop_control.htm).

## Recap
&lt;center&gt;![](../images/recap.webp){height=300px}&lt;/center&gt;
Loops are really useful when we need to perform the same task, or function, over-and-over again. Where we have data stored in a data structure such
as a list, we can apply something to everything in the list using a for loop.
This is an efficient way to code as it means we do not need to duplicate the instructions.

&lt;!--chapter:end:07_loops.Rmd--&gt;

---
title: &quot;Python for Bioinformatics&quot;
author: Graeme Fox
date: &quot;September, 2022&quot;
---
# Reading sequence data

&lt;center&gt;![](../images/DNA_icon.png){height=300px}![](../images/biopython_logo.png){height=300px}&lt;/center&gt;

Biopython is a suite of libraries and tools designed for analysis of biological datasets in Python, largely centered around the processing of DNA sequence data.

You can find much more information about Biopython on the [website](https://biopython.org/).

During this section on Biopython we are going to be using the SeqIO module. The knowledge and approaches from the previous chapters
will now be used to process some genomic DNA sequence data.

## SeqIO

&lt;center&gt;![](../images/DNA_computer.png){height=300px}&lt;/center&gt;

The SeqIO tools are the main Sequence Input/Output interface for Biopython. It provides a relatively simple, uniform interface for working with various types of DNA sequence files. The module reads raw sequence files (such as .FASTQ files you will likely receive from your sequencing facility) and converts them to a Python readable format; a [SeqRecord] (https://biopython.org/wiki/SeqRecord) class.


## Reading a FASTQ DNA data file into Python
One of the most common types of DNA sequence data file you will be working with is the [FASTQ] (https://en.wikipedia.org/wiki/FASTQ_format) file. 
The FASTQ file contains multiple DNA sequences, each one might be a single short read from an Illumina sequencer and the total file can contain thousands or millions of sequences depending on which sequencer was used to produce it.

The FASTQ has a very specific format and holds three important pieces of information.

## The FASTQ format

&lt;center&gt;![](../images/fastq_icon.png){height=300px}&lt;/center&gt;


The file has a repetitive structure of four distinct components:
* A unique sequence identifier (ID)
* The DNA nucleotide sequence
* A small spacer
* The quality score associated with each nucleotide

The ID line always starts with the @ symbol and this is followed by a string of numbers and letters giving details such as the sequencer that produced the data, the run number, an index adapter used on the sample.

The DNA nucleotide sequence is shown as a string of ATGC characters. 

There is a spacer consisting of the &quot;+&quot; character which separates the sequence from the quality scores.

The quality score associated with each nucleotide above is ASCII encoded with each symbol representing a quality score from 0 to 40 (higher being better). This line is not meant to be easily human-readable.

An example of a single unit of this four line repeat can be seen below:</code></pre>
<p><span class="citation">@MISEQ:42:000000000-A8DD3:1:1101:15046:1468</span>
TTCACATCCTGACCATTCAGTTGAGCAAAATAGTTCTTCAGT
+
CCFFFFFGHHHHJIJJJJIJJJIIJJJJIIIJJGFIIIJEDD</p>
<pre><code>## Read a FASTQ file into Python

&lt;br&gt; Scenario: &lt;/b&gt;You have just received your FASTQ data back from your sequencing facility. This was a paired-end sequencing run meaning that we have a forward (R1) read file and a reverse (R2) read file; both are in the FASTQ format. The data was generated using a shotgun, whole-genome sequencing approach to sequence the genome of the undulate ray (*Raja undulata*).

&lt;center&gt;![](../images/undulate_ray.jpeg){height=300px}&lt;/center&gt;

We are going to read the data into Python ready to do some data processing and analyses.

## Setting up a script to use BioPython

&lt;center&gt;![](../images/vscode_icon.jpeg){height=300px}&lt;/center&gt;

Open a new Python script in VScode (File &gt; New Text File).

We need to load a couple of Python libraries: 

&#39;Biopython&#39; (a suite of bioinformatics focused Python tools), and 

&#39;re&#39; (allows us to perform searches using wildcards and special characters). 

Add the following to the top of the script.
</code></pre>
<p>from Bio import SeqIO
import re</p>
<pre><code>
You should hopefully still have the current working directory on your clipboard.
Set a variable called &#39;wd&#39;, (working directory) to the filepath on your clipboard by adding a version of the following:
</code></pre>
<p>wd = “/pub39/tea/nsc014”</p>
<pre><code>
## Read a fie of sequence data

&lt;center&gt;![](../images/bioinformatics_icon2.png){height=300px}&lt;/center&gt;

We are going to use &#39;SeqIO&#39; (which stands for sequence in/out) and the &#39;parse&#39; function which splits the FASTQ file into its separate components.

We store the information on which file to parse in the variable called &#39;R_undulata_R1_data. SeqIO.parse requires two arguments: the filepath to the file which is to be read, and the format of that file.
</code></pre>
<p>R_undulata_R1_data = SeqIO.parse(wd + “data/Raja_undulata_WGS_trim_R1.fastq”, “fastq”)</p>
<pre><code>
We can then look through the file using a *for* loop. Each of the four-unit repeats of the FASTQ file is read as a separate &#39;record&#39; and we can access them in the script. 

To access the ID line of each record, we use &#39;.id&#39; and to access the sequence we use &#39;.seq&#39;. Each of these elements can be printed in the terminal using an appropriate print statement.
</code></pre>
<p>for record in R_undulata_R1_data:
print(record.id)
print(record.seq)</p>
<pre><code>
We can perform operations on each of these elements. For example, Biopython has a built in translate() method to convert nucleotide sequence into protein sequence (although bear in mind that we are not considering reading frames and we likely have lots of non-coding data).

Modify the for loop above to the following:</code></pre>
<p>for record in R_undulata_R1_data:
#print(record.id)
#print(record.seq)
print(record.seq.translate())</p>
<pre><code>
We can implement a counter to tally the total number of reads in the file. Modify the *for* loop above to the following (you do not need to add the comments however):
</code></pre>
<div id="start-a-counter-to-track-how-many-reads" class="section level2 hasAnchor" number="15.1">
<h2><span class="header-section-number">15.1</span> start a counter to track how many reads<a href="begin-the-while-loop.html#start-a-counter-to-track-how-many-reads" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>counter = 0</p>
<p>for record in R_undulata_R1_data:
#print(record.id)
#print(record.seq)
#print(record.seq.translate())
# increase the counter by 1 each time it sees a new read
counter = counter + 1</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="initialise-a-variable-to-add-the-daily-temps-together.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="save-the-number-of-raw-sequences-for-use-later.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
