<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>10 Reading sequence data | Python for Bioinformatics</title>
  <meta name="description" content="NEOF book for the Python for Bioinformatics workshop" />
  <meta name="generator" content="bookdown 0.29 and GitBook 2.6.7" />

  <meta property="og:title" content="10 Reading sequence data | Python for Bioinformatics" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="/images/NEOF.png" />
  <meta property="og:description" content="NEOF book for the Python for Bioinformatics workshop" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="10 Reading sequence data | Python for Bioinformatics" />
  
  <meta name="twitter:description" content="NEOF book for the Python for Bioinformatics workshop" />
  <meta name="twitter:image" content="/images/NEOF.png" />

<meta name="author" content="Graeme Fox" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="images/NEOF_favicon.png" type="image/x-icon" />
<link rel="prev" href="loops.html"/>
<link rel="next" href="using-conditions-to-modify-dna-data.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>




</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="python-for-bioinformatics.html"><a href="python-for-bioinformatics.html"><i class="fa fa-check"></i><b>2</b> Python for bioinformatics</a>
<ul>
<li class="chapter" data-level="2.1" data-path="python-for-bioinformatics.html"><a href="python-for-bioinformatics.html#materials-covered-in-the-course"><i class="fa fa-check"></i><b>2.1</b> Materials Covered in the Course</a></li>
<li class="chapter" data-level="2.2" data-path="python-for-bioinformatics.html"><a href="python-for-bioinformatics.html#inputting-commands"><i class="fa fa-check"></i><b>2.2</b> Inputting Commands</a></li>
<li class="chapter" data-level="2.3" data-path="python-for-bioinformatics.html"><a href="python-for-bioinformatics.html#pseudocode"><i class="fa fa-check"></i><b>2.3</b> Pseudocode</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="cluster-introduction.html"><a href="cluster-introduction.html"><i class="fa fa-check"></i><b>3</b> Cluster Introduction</a>
<ul>
<li class="chapter" data-level="3.1" data-path="cluster-introduction.html"><a href="cluster-introduction.html#logon-instructions"><i class="fa fa-check"></i><b>3.1</b> Logon Instructions</a></li>
<li class="chapter" data-level="3.2" data-path="cluster-introduction.html"><a href="cluster-introduction.html#the-terminal-window"><i class="fa fa-check"></i><b>3.2</b> The Terminal Window</a></li>
<li class="chapter" data-level="3.3" data-path="cluster-introduction.html"><a href="cluster-introduction.html#load-workshop-data"><i class="fa fa-check"></i><b>3.3</b> Load workshop data</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="scripting-in-python.html"><a href="scripting-in-python.html"><i class="fa fa-check"></i><b>4</b> Scripting in Python</a>
<ul>
<li class="chapter" data-level="4.1" data-path="scripting-in-python.html"><a href="scripting-in-python.html#introduction-to-commands-and-scripts"><i class="fa fa-check"></i><b>4.1</b> Introduction to commands and scripts</a></li>
<li class="chapter" data-level="4.2" data-path="scripting-in-python.html"><a href="scripting-in-python.html#issuing-commands-directly-to-the-interpretator"><i class="fa fa-check"></i><b>4.2</b> Issuing commands directly to the interpretator</a></li>
<li class="chapter" data-level="4.3" data-path="scripting-in-python.html"><a href="scripting-in-python.html#a-note-on-indentation"><i class="fa fa-check"></i><b>4.3</b> A note on indentation</a></li>
<li class="chapter" data-level="4.4" data-path="scripting-in-python.html"><a href="scripting-in-python.html#tabs-vs-spaces"><i class="fa fa-check"></i><b>4.4</b> Tabs vs spaces</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="python-fundamentals.html"><a href="python-fundamentals.html"><i class="fa fa-check"></i><b>5</b> Python Fundamentals</a>
<ul>
<li class="chapter" data-level="5.1" data-path="python-fundamentals.html"><a href="python-fundamentals.html#operators"><i class="fa fa-check"></i><b>5.1</b> Operators</a></li>
<li class="chapter" data-level="5.2" data-path="python-fundamentals.html"><a href="python-fundamentals.html#comments"><i class="fa fa-check"></i><b>5.2</b> Comments</a></li>
<li class="chapter" data-level="5.3" data-path="python-fundamentals.html"><a href="python-fundamentals.html#variables"><i class="fa fa-check"></i><b>5.3</b> Variables</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="python-fundamentals.html"><a href="python-fundamentals.html#strings"><i class="fa fa-check"></i><b>5.3.1</b> Strings</a></li>
<li class="chapter" data-level="5.3.2" data-path="python-fundamentals.html"><a href="python-fundamentals.html#integers"><i class="fa fa-check"></i><b>5.3.2</b> Integers</a></li>
<li class="chapter" data-level="5.3.3" data-path="python-fundamentals.html"><a href="python-fundamentals.html#floats-floating-point-numbers"><i class="fa fa-check"></i><b>5.3.3</b> Floats (floating point numbers)</a></li>
<li class="chapter" data-level="5.3.4" data-path="python-fundamentals.html"><a href="python-fundamentals.html#booleans"><i class="fa fa-check"></i><b>5.3.4</b> Booleans</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="python-fundamentals.html"><a href="python-fundamentals.html#classes-of-variable"><i class="fa fa-check"></i><b>5.4</b> Classes of variable</a></li>
<li class="chapter" data-level="5.5" data-path="python-fundamentals.html"><a href="python-fundamentals.html#transforming-variables"><i class="fa fa-check"></i><b>5.5</b> Transforming variables</a></li>
<li class="chapter" data-level="5.6" data-path="python-fundamentals.html"><a href="python-fundamentals.html#functions-and-methods"><i class="fa fa-check"></i><b>5.6</b> Functions and methods</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="python-fundamentals.html"><a href="python-fundamentals.html#running-a-function"><i class="fa fa-check"></i><b>5.6.1</b> Running a function</a></li>
<li class="chapter" data-level="5.6.2" data-path="python-fundamentals.html"><a href="python-fundamentals.html#running-a-method"><i class="fa fa-check"></i><b>5.6.2</b> Running a method</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="python-fundamentals.html"><a href="python-fundamentals.html#recap"><i class="fa fa-check"></i><b>5.7</b> Recap</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="data-structures.html"><a href="data-structures.html"><i class="fa fa-check"></i><b>6</b> Data Structures</a>
<ul>
<li class="chapter" data-level="6.1" data-path="data-structures.html"><a href="data-structures.html#lists"><i class="fa fa-check"></i><b>6.1</b> Lists</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="data-structures.html"><a href="data-structures.html#add-numeric-data-to-a-list"><i class="fa fa-check"></i><b>6.1.1</b> Add numeric data to a list</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="data-structures.html"><a href="data-structures.html#subsetting-data"><i class="fa fa-check"></i><b>6.2</b> Subsetting data</a></li>
<li class="chapter" data-level="6.3" data-path="data-structures.html"><a href="data-structures.html#set"><i class="fa fa-check"></i><b>6.3</b> Set</a></li>
<li class="chapter" data-level="6.4" data-path="data-structures.html"><a href="data-structures.html#transform-a-list-to-a-set"><i class="fa fa-check"></i><b>6.4</b> Transform a list to a set</a></li>
<li class="chapter" data-level="6.5" data-path="data-structures.html"><a href="data-structures.html#dictionary"><i class="fa fa-check"></i><b>6.5</b> Dictionary</a></li>
<li class="chapter" data-level="6.6" data-path="data-structures.html"><a href="data-structures.html#recap-1"><i class="fa fa-check"></i><b>6.6</b> Recap</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="python-conditions-if-and-else-statements.html"><a href="python-conditions-if-and-else-statements.html"><i class="fa fa-check"></i><b>7</b> Python Conditions: ‘if’ and ‘else’ statements</a>
<ul>
<li class="chapter" data-level="7.1" data-path="python-conditions-if-and-else-statements.html"><a href="python-conditions-if-and-else-statements.html#if-statements"><i class="fa fa-check"></i><b>7.1</b> If statements</a></li>
<li class="chapter" data-level="7.2" data-path="python-conditions-if-and-else-statements.html"><a href="python-conditions-if-and-else-statements.html#if-else-statements"><i class="fa fa-check"></i><b>7.2</b> If else statements</a></li>
<li class="chapter" data-level="7.3" data-path="python-conditions-if-and-else-statements.html"><a href="python-conditions-if-and-else-statements.html#recap-2"><i class="fa fa-check"></i><b>7.3</b> Recap</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="starting-scripting-in-python.html"><a href="starting-scripting-in-python.html"><i class="fa fa-check"></i><b>8</b> Starting scripting in Python</a>
<ul>
<li class="chapter" data-level="8.1" data-path="starting-scripting-in-python.html"><a href="starting-scripting-in-python.html#open-a-plain-text-editor"><i class="fa fa-check"></i><b>8.1</b> Open a plain text editor</a></li>
<li class="chapter" data-level="8.2" data-path="starting-scripting-in-python.html"><a href="starting-scripting-in-python.html#running-a-script-from-the-terminal"><i class="fa fa-check"></i><b>8.2</b> Running a script from the terminal</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="loops.html"><a href="loops.html"><i class="fa fa-check"></i><b>9</b> Loops</a>
<ul>
<li class="chapter" data-level="9.1" data-path="loops.html"><a href="loops.html#data-structure-recap"><i class="fa fa-check"></i><b>9.1</b> Data Structure Recap</a></li>
<li class="chapter" data-level="9.2" data-path="loops.html"><a href="loops.html#introduction-to-loops"><i class="fa fa-check"></i><b>9.2</b> Introduction To Loops</a></li>
<li class="chapter" data-level="9.3" data-path="loops.html"><a href="loops.html#for-loops"><i class="fa fa-check"></i><b>9.3</b> For Loops</a></li>
<li class="chapter" data-level="9.4" data-path="loops.html"><a href="loops.html#for-loop-example"><i class="fa fa-check"></i><b>9.4</b> For Loop Example</a></li>
<li class="chapter" data-level="9.5" data-path="loops.html"><a href="loops.html#use-a-for-to-complement-a-dna-sequence"><i class="fa fa-check"></i><b>9.5</b> Use a for to complement a DNA sequence</a></li>
<li class="chapter" data-level="9.6" data-path="loops.html"><a href="loops.html#use-a-for-to-reverse-complement-a-dna-sequence"><i class="fa fa-check"></i><b>9.6</b> Use a for to reverse-complement a DNA sequence</a></li>
<li class="chapter" data-level="9.7" data-path="loops.html"><a href="loops.html#while-loops"><i class="fa fa-check"></i><b>9.7</b> While loops</a></li>
<li class="chapter" data-level="9.8" data-path="loops.html"><a href="loops.html#further-control-statements"><i class="fa fa-check"></i><b>9.8</b> Further control statements</a></li>
<li class="chapter" data-level="9.9" data-path="loops.html"><a href="loops.html#recap-3"><i class="fa fa-check"></i><b>9.9</b> Recap</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="reading-sequence-data.html"><a href="reading-sequence-data.html"><i class="fa fa-check"></i><b>10</b> Reading sequence data</a>
<ul>
<li class="chapter" data-level="10.1" data-path="reading-sequence-data.html"><a href="reading-sequence-data.html#seqio"><i class="fa fa-check"></i><b>10.1</b> SeqIO</a></li>
<li class="chapter" data-level="10.2" data-path="reading-sequence-data.html"><a href="reading-sequence-data.html#reading-a-fastq-dna-data-file-into-python"><i class="fa fa-check"></i><b>10.2</b> Reading a FASTQ DNA data file into Python</a></li>
<li class="chapter" data-level="10.3" data-path="reading-sequence-data.html"><a href="reading-sequence-data.html#the-fastq-format"><i class="fa fa-check"></i><b>10.3</b> The FASTQ format</a></li>
<li class="chapter" data-level="10.4" data-path="reading-sequence-data.html"><a href="reading-sequence-data.html#read-a-fastq-file-into-python"><i class="fa fa-check"></i><b>10.4</b> Read a FASTQ file into Python</a></li>
<li class="chapter" data-level="10.5" data-path="reading-sequence-data.html"><a href="reading-sequence-data.html#setting-up-a-script-to-use-biopython"><i class="fa fa-check"></i><b>10.5</b> Setting up a script to use BioPython</a></li>
<li class="chapter" data-level="10.6" data-path="reading-sequence-data.html"><a href="reading-sequence-data.html#read-a-fie-of-sequence-data"><i class="fa fa-check"></i><b>10.6</b> Read a fie of sequence data</a></li>
<li class="chapter" data-level="10.7" data-path="reading-sequence-data.html"><a href="reading-sequence-data.html#modifying-reads"><i class="fa fa-check"></i><b>10.7</b> Modifying reads</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="using-conditions-to-modify-dna-data.html"><a href="using-conditions-to-modify-dna-data.html"><i class="fa fa-check"></i><b>11</b> Using Conditions to Modify DNA Data</a>
<ul>
<li class="chapter" data-level="11.1" data-path="using-conditions-to-modify-dna-data.html"><a href="using-conditions-to-modify-dna-data.html#remove-short-reads"><i class="fa fa-check"></i><b>11.1</b> Remove short reads</a></li>
<li class="chapter" data-level="11.2" data-path="using-conditions-to-modify-dna-data.html"><a href="using-conditions-to-modify-dna-data.html#remove-pcr-primers"><i class="fa fa-check"></i><b>11.2</b> Remove PCR primers</a></li>
<li class="chapter" data-level="11.3" data-path="using-conditions-to-modify-dna-data.html"><a href="using-conditions-to-modify-dna-data.html#recap-4"><i class="fa fa-check"></i><b>11.3</b> Recap</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="reading-and-writing-files.html"><a href="reading-and-writing-files.html"><i class="fa fa-check"></i><b>12</b> Reading and writing files</a>
<ul>
<li class="chapter" data-level="12.1" data-path="reading-and-writing-files.html"><a href="reading-and-writing-files.html#open"><i class="fa fa-check"></i><b>12.1</b> ‘open’</a></li>
<li class="chapter" data-level="12.2" data-path="reading-and-writing-files.html"><a href="reading-and-writing-files.html#with"><i class="fa fa-check"></i><b>12.2</b> ‘with’</a></li>
<li class="chapter" data-level="12.3" data-path="reading-and-writing-files.html"><a href="reading-and-writing-files.html#read-a-data-file"><i class="fa fa-check"></i><b>12.3</b> Read a data file</a></li>
<li class="chapter" data-level="12.4" data-path="reading-and-writing-files.html"><a href="reading-and-writing-files.html#write-data-to-a-file"><i class="fa fa-check"></i><b>12.4</b> Write data to a file</a></li>
<li class="chapter" data-level="12.5" data-path="reading-and-writing-files.html"><a href="reading-and-writing-files.html#write-out-the-filtered-undulate-ray-sequence-data"><i class="fa fa-check"></i><b>12.5</b> Write out the filtered Undulate ray sequence data</a></li>
<li class="chapter" data-level="12.6" data-path="reading-and-writing-files.html"><a href="reading-and-writing-files.html#recap-5"><i class="fa fa-check"></i><b>12.6</b> Recap</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Python for Bioinformatics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="reading-sequence-data" class="section level1 hasAnchor" number="10">
<h1><span class="header-section-number">10</span> Reading sequence data<a href="reading-sequence-data.html#reading-sequence-data" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<center>
<img src="../images/DNA_icon.png" height="300" /><img src="../images/biopython_logo.png" height="300" />
</center>
<p>Biopython is a suite of libraries and tools designed for analysis of biological datasets in Python, largely centered around the processing of DNA sequence data.</p>
<p>You can find much more information about Biopython on the <a href="https://biopython.org/">website</a>.</p>
<p>During this section on Biopython we are going to be using the SeqIO module. The knowledge and approaches from the previous chapters
will now be used to process some genomic DNA sequence data.</p>
<div id="seqio" class="section level2 hasAnchor" number="10.1">
<h2><span class="header-section-number">10.1</span> SeqIO<a href="reading-sequence-data.html#seqio" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="../images/DNA_computer.png" height="300" />
</center>
<p>The SeqIO tools are the main Sequence Input/Output interface for Biopython. It provides a relatively simple, uniform interface for working with various types of DNA sequence files. The module reads raw sequence files (such as .FASTQ files you will likely receive from your sequencing facility) and converts them to a Python readable format; a <a href="https://biopython.org/wiki/SeqRecord">SeqRecord</a> class.</p>
</div>
<div id="reading-a-fastq-dna-data-file-into-python" class="section level2 hasAnchor" number="10.2">
<h2><span class="header-section-number">10.2</span> Reading a FASTQ DNA data file into Python<a href="reading-sequence-data.html#reading-a-fastq-dna-data-file-into-python" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>One of the most common types of DNA sequence data file you will be working with is the [FASTQ] (<a href="https://en.wikipedia.org/wiki/FASTQ_format" class="uri">https://en.wikipedia.org/wiki/FASTQ_format</a>) file.
The FASTQ file contains multiple DNA sequences, each one might be a single short read from an Illumina sequencer and the total file can contain thousands or millions of sequences depending on which sequencer was used to produce it.</p>
<p>The FASTQ has a very specific format and holds three important pieces of information.</p>
</div>
<div id="the-fastq-format" class="section level2 hasAnchor" number="10.3">
<h2><span class="header-section-number">10.3</span> The FASTQ format<a href="reading-sequence-data.html#the-fastq-format" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="../images/fastq_icon.png" height="300" />
</center>
<p>The file has a repetitive structure of four distinct components:
* A unique sequence identifier (ID)
* The DNA nucleotide sequence
* A small spacer
* The quality score associated with each nucleotide</p>
<p>The ID line always starts with the @ symbol and this is followed by a string of numbers and letters giving details such as the sequencer that produced the data, the run number, an index adapter used on the sample.</p>
<p>The DNA nucleotide sequence is shown as a string of ATGC characters.</p>
<p>There is a spacer consisting of the “+” character which separates the sequence from the quality scores.</p>
<p>The quality score associated with each nucleotide above is ASCII encoded with each symbol representing a quality score from 0 to 40 (higher being better). This line is not meant to be easily human-readable.</p>
<p>An example of a single unit of this four line repeat can be seen below:</p>
<pre><code>@MISEQ:42:000000000-A8DD3:1:1101:15046:1468
TTCACATCCTGACCATTCAGTTGAGCAAAATAGTTCTTCAGT
+
CCFFFFFGHHHHJIJJJJIJJJIIJJJJIIIJJGFIIIJEDD</code></pre>
</div>
<div id="read-a-fastq-file-into-python" class="section level2 hasAnchor" number="10.4">
<h2><span class="header-section-number">10.4</span> Read a FASTQ file into Python<a href="reading-sequence-data.html#read-a-fastq-file-into-python" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><br> Scenario: </b>You have just some FASTQ formatted data back from your sequencing facility. The data was generated using a shotgun, whole-genome sequencing approach to sequence the genome of the undulate ray (<em>Raja undulata</em>).</p>
<center>
<img src="../images/undulate_ray.jpeg" height="300" />
</center>
<p>We are going to read the data into Python ready to do some data processing and analyses.</p>
</div>
<div id="setting-up-a-script-to-use-biopython" class="section level2 hasAnchor" number="10.5">
<h2><span class="header-section-number">10.5</span> Setting up a script to use BioPython<a href="reading-sequence-data.html#setting-up-a-script-to-use-biopython" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Open a new file in Gedit (File &gt; New).</p>
<p>We need to load a couple of Python libraries:</p>
<p>‘Biopython’ (a suite of bioinformatics focused Python tools)</p>
<p>…and…</p>
<p>‘re’ (allows us to perform searches using wildcards and special characters).</p>
<p>Add the following to the top of the script.</p>
<pre><code>from Bio import SeqIO
import re</code></pre>
<p>You should hopefully still have the current working directory on your clipboard.
Set a variable called ‘wd’, (working directory) to the filepath on your clipboard by adding a version of the following:</p>
<pre><code>wd = &quot;/pub39/tea/nsc014&quot;</code></pre>
</div>
<div id="read-a-fie-of-sequence-data" class="section level2 hasAnchor" number="10.6">
<h2><span class="header-section-number">10.6</span> Read a fie of sequence data<a href="reading-sequence-data.html#read-a-fie-of-sequence-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<center>
<img src="../images/bioinformatics_icon2.png" height="300" />
</center>
<p>We are going to use ‘SeqIO’ (which stands for sequence in/out) and the ‘parse’ function which splits the FASTQ file into its separate components.</p>
<p>We store the information on which file to parse in the variable called ’R_undulata_R1_data. SeqIO.parse requires two arguments: the filepath to the file which is to be read, and the format of that file.</p>
<pre><code>R_undulata_R1_data = SeqIO.parse(wd + &quot;data/Raja_undulata_WGS_trim_R1.fastq&quot;, &quot;fastq&quot;)</code></pre>
<p>We can then look through the file using a <b>for</b> loop. Each of the four-unit repeats of the FASTQ file is read as a separate ‘record’ and we can access them in the script.</p>
<p>To access the ID line of each record, we use ‘.id’ and to access the sequence we use ‘.seq’. Each of these elements can be printed in the terminal using an appropriate print statement.</p>
<pre><code>for record in R_undulata_R1_data:
  print(record.id)
  print(record.seq)</code></pre>
<p>Save the new script as ‘script02.py’ and run it from the terminal window to print the ID and sequences.</p>
</div>
<div id="modifying-reads" class="section level2 hasAnchor" number="10.7">
<h2><span class="header-section-number">10.7</span> Modifying reads<a href="reading-sequence-data.html#modifying-reads" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We can perform operations on each of these elements. For example, Biopython has a built in translate() method to convert nucleotide sequence into protein sequence (although bear in mind that we are not considering reading frames and we likely have lots of non-coding data).</p>
<p>Modify the for loop above to the following:</p>
<pre><code>for record in R_undulata_R1_data:
  #print(record.id)
  #print(record.seq)
  print(record.seq.translate())</code></pre>
<p>We can implement a counter to tally the total number of reads in the file. Modify the <b>for</b> loop above to the following (you do not need to add the comments however):</p>
<pre><code>## start a counter to track how many reads
counter = 0

for record in R_undulata_R1_data:
  #print(record.id) 
  #print(record.seq)
  #print(record.seq.translate())
  # increase the counter by 1 each time it sees a new read
  counter = counter + 1

# save the number of raw sequences for use later
number_raw_reads = counter

print(&quot;Raw sequence file has &quot; + str(number_raw_reads) + \
&quot; sequence records.&quot;)</code></pre>
<p>Note that we need to convert the number_raw_reads to a string type in order for the print statement to work using str(). Python will not print variables of different types in the same statement.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="loops.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="using-conditions-to-modify-dna-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
